type-definitions/immutable.js.flow

```diff
@@ -43,9 +58,15 @@ declare class _Collection<K, +V> /*implements ValueObject*/ {
   first(): V | void;
   last(): V | void;
 
-  getIn(keyPath: Iterable<mixed>, notSetValue?: mixed): any;
   hasIn(keyPath: Iterable<mixed>): boolean;
 
+  getIn(keyPath: [], notSetValue?: mixed): this;
+  getIn<NSV>(keyPath: [K], notSetValue: NSV): V | NSV;
+  getIn<NSV, K2: $KeyOf<V>>(keyPath: [K, K2], notSetValue: NSV): $ValOf<V, K2> | NSV;
+  getIn<NSV, K2: $KeyOf<V>, K3: $KeyOf<$ValOf<V, K2>>>(keyPath: [K, K2, K3], notSetValue: NSV): $ValOf<$ValOf<V, K2>, K3> | NSV;
+  getIn<NSV, K2: $KeyOf<V>, K3: $KeyOf<$ValOf<V, K2>>, K4: $KeyOf<$ValOf<$ValOf<V, K2>, K3>>>(keyPath: [K, K2, K3, K4], notSetValue: NSV): $ValOf<$ValOf<$ValOf<V, K2>, K3>, K4> | NSV;
+  getIn<NSV, K2: $KeyOf<V>, K3: $KeyOf<$ValOf<V, K2>>, K4: $KeyOf<$ValOf<$ValOf<V, K2>, K3>>, K5: $KeyOf<$ValOf<$ValOf<$ValOf<V, K2>, K3>, K4>>>(keyPath: [K, K2, K3, K4, K5], notSetValue: NSV): $ValOf<$ValOf<$ValOf<$ValOf<V, K2>, K3>, K4>, K5> | NSV;
+
```

Comment:
One thing I'm not sure about is not having a case for supporting keyPaths with an unknown or more than 5 elements. Adding back `getIn(keyPath: Iterable<mixed>, notSetValue?: mixed): any;` causes some of the `$ExpectError` to no longer trigger when checking for valid key paths.

Reply:
I’ve been wanting to introduce a `$Reduce<>` utility type. Such a type would allow you to support key paths of any length 😊. The definition of `getIn()` with `$Reduce<>` may look something like:

```
getIn<Path, NSV>(
  keyPath: Path,
  notSetValue: NSV,
): $Reduce<Path, <Acc, K>(Acc, K) => $ValOf<Acc, K>, this> | NSV;
```